src/game/client/draw/w3dlaserdraw.cpp

```diff
@@ -13,3 +13,314 @@
+                alpha = inneralpha;
+                red = innerred * inneralpha;
+                green = innergreen * inneralpha;
+                blue = innerblue * inneralpha;
+            } else {
+                int i1 = j / (data->m_numBeams - 1.0);
+                beamwidth = (data->m_outerBeamWidth - data->m_innerBeamWidth) * i1 + data->m_innerBeamWidth;
+                alpha = (outeralpha - inneralpha) * i1 + inneralpha;
+                red = (outerred - innerred) * i1 * inneralpha + innerred;
+                green = (outergreen - innergreen) * i1 * inneralpha + innergreen;
+                blue = (outerblue - innerblue) * i1 * inneralpha + innerblue;
+            }
+
+            m_line3D[index] = new SegmentedLineClass();
+            SegmentedLineClass *line = m_line3D[index];
+
+            if (line) {
+                line->Set_Texture(m_texture);
+                line->Set_Shader(ShaderClass::s_presetAdditiveShader);
+                line->Set_Width(beamwidth);
+                line->Set_Color(Vector3(red, green, blue));
+                line->Set_UV_Offset_Rate(Vector2(0.0f, data->m_scrollRate));
+
+                if (m_texture) {
+                    line->Set_Texture_Mapping_Mode(SegLineRendererClass::TILED_TEXTURE_MAP);
+                }
+
+                W3DDisplay::s_3DScene->Add_Render_Object(line);
+                line->Set_Visible(0);
+            }
+        }
+    }
+}
+
+W3DLaserDraw::~W3DLaserDraw()
+{
+    const W3DLaserDrawModuleData *data = Get_Module_Data();
+
+    for (unsigned int i = 0; i < data->m_segments * data->m_numBeams; i++) {
+        W3DDisplay::s_3DScene->Remove_Render_Object(m_line3D[i]);
+        Ref_Ptr_Release(m_line3D[i]);
+    }
+
+    delete[] m_line3D;
+}
+
+float W3DLaserDraw::Get_Laser_Template_Width() const
+{
+    return Get_Module_Data()->m_outerBeamWidth * 0.5f;
+}
+
+ModuleData *W3DLaserDraw::Friend_New_Module_Data(INI *ini)
+{
+    W3DLaserDrawModuleData *data = new W3DLaserDrawModuleData();
+
+    if (ini) {
+        ini->Init_From_INI_Multi_Proc(data, W3DLaserDrawModuleData::Build_Field_Parse);
+    }
+
+    return data;
+}
+
+Module *W3DLaserDraw::Friend_New_Module_Instance(Thing *thing, ModuleData const *module_data)
+{
+    return NEW_POOL_OBJ(W3DLaserDraw, thing, module_data);
+}
+
+NameKeyType W3DLaserDraw::Get_Module_Name_Key() const
+{
+    static NameKeyType nk = g_theNameKeyGenerator->Name_To_Key("W3DLaserDraw");
+    return nk;
+}
+
+void W3DLaserDraw::Do_Draw_Module(Marix3D *transform)
+{
+    const W3DLaserDrawModuleData *data = Get_Module_Data();
+    Drawable *drawable = Get_Drawable();
+    static NameKeyType key_LaserUpdate = g_theNameKeyGenerator->Name_To_Key("LaserUpdate");
+    LaserUpdate *update = (LaserUpdate *)drawable->Find_Client_Update_Module(key_LaserUpdate);
+
+    if (update) {
+        if (update->Is_Dirty() || m_setLaserPosition) {
+            update->Set_Dirty(false);
+            m_setLaserPosition = false;
+            Vector3 points[2];

```

Comment:
It looks as if `points` could remain uninitialized when used later.

Reply:
Not possible.
No matter which way the if (data->m_arcHeight <= 0.0f || data->m_segments <= 1) goes, points will be set.